truffle/src/main/java/org/jruby/truffle/runtime/rope/RopeOperations.java

```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright (c) 2016 Oracle and/or its affiliates. All rights reserved. This
+ * code is released under a tri EPL/GPL/LGPL license. You can use it,
+ * redistribute it and/or modify it under the terms of the:
+ *
+ * Eclipse Public License version 1.0
+ * GNU General Public License version 2
+ * GNU Lesser General Public License version 2.1
+ */
+package org.jruby.truffle.runtime.rope;
+
+import com.oracle.truffle.api.CompilerDirectives;
+import org.jcodings.Encoding;
+import org.jcodings.specific.ASCIIEncoding;
+import org.jcodings.specific.UTF8Encoding;
+import org.jruby.RubyEncoding;
+import org.jruby.util.StringSupport;
+import org.jruby.util.io.EncodingUtils;
+
+public class RopeOperations {
+
+    public static final Rope EMPTY_ASCII_8BIT_ROPE = create(new byte[] {}, ASCIIEncoding.INSTANCE, StringSupport.CR_7BIT);
+    public static final Rope EMPTY_UTF8_ROPE = create(new byte[] {}, UTF8Encoding.INSTANCE, StringSupport.CR_7BIT);
+
+    public static LeafRope create(byte[] bytes, Encoding encoding, int codeRange) {
+        int characterLength = -1;
+
+        if (codeRange == StringSupport.CR_UNKNOWN) {
+            final long packedLengthAndCodeRange = calculateCodeRangeAndLength(encoding, bytes, 0, bytes.length);
+
+            codeRange = StringSupport.unpackArg(packedLengthAndCodeRange);
+            characterLength = StringSupport.unpackResult(packedLengthAndCodeRange);
+        } else if (codeRange == StringSupport.CR_VALID) {
+            characterLength = strLength(encoding, bytes, 0, bytes.length);
+        }
+
+        switch(codeRange) {
+            case StringSupport.CR_7BIT: return new AsciiOnlyLeafRope(bytes, encoding);
+            case StringSupport.CR_VALID: return new ValidLeafRope(bytes, encoding, characterLength);
+            case StringSupport.CR_BROKEN: return new InvalidLeafRope(bytes, encoding);
+            default: {
+                CompilerDirectives.transferToInterpreter();
+                throw new RuntimeException(String.format("Unknown code range type: %d", codeRange));
+            }
+        }
+    }
+
+    @CompilerDirectives.TruffleBoundary
+    public static Rope concat(Rope left, Rope right, Encoding encoding) {
+        if (right.isEmpty()) {
+            return withEncoding(left, encoding);
+        }
+
+        if (left.isEmpty()) {
+            return withEncoding(right, encoding);
+        }
+
+        return new ConcatRope(left, right, encoding);
+    }
+
+    public static Rope substring(Rope base, int offset, int byteLength) {
+        if (byteLength == 0) {
+            return withEncoding(EMPTY_UTF8_ROPE, base.getEncoding());
+        }
+
+        if (byteLength - offset == base.byteLength()) {
+            return base;
+        }
+
+        if (base instanceof SubstringRope) {
```

Comment:
Yes, but we also need to be able to create ropes outside the context of Ruby execution. I agree the `instanceof` are bad here. I plan on tidying this up as follow-on work.